repo/pull_repo.py

```diff
@@ -0,0 +1,128 @@
+import requests
+import json
+from datetime import datetime
+import os
+
+
+class EmptyResponse(Exception):
+    pass
+
+
+class ServiceUnavailable(Exception):
+    pass
+
+
+class ForbiddenException(Exception):
+    pass
+
+
+batch = {}
+
+
+def parse_link(string):
+    data = string.split()
+    next_link = data[0][1:-2]
+
+    if "last" in string:
+        return next_link, True
+    else:
+        return "", False
+
+
+def get_token():
+    token = os.environ.get("GITHUB_TOKEN")
+    if not token:
+        raise Exception("No token in env")
+
+    return token
+
+
+def pull_data(url, etag=None):
+    # User access token requests are subject to a higher limit of 15,000 requests per hour
+    headers = {"Authorization": f"Bearer {get_token()}"}
+
+    if etag:
+        headers["ETag"] = etag
+
+    data = requests.get(
+        url,
+        params={"per_page": 100},
+        headers=headers,
+    )
+    next_link = ""
+    f = False
+
+    if data.status_code == 200:
+        etag = data.headers["ETag"]
+
+        if "Link" in data.headers.keys():
+            next_link, f = parse_link(data.headers["Link"])
+
+        return data.text, next_link, f, etag
+
+    if data.status_code == 304:
+        raise EmptyResponse
+    elif data.status_code == 403:
```

Comment:
we need to respect the poll interval header https://docs.github.com/en/rest/activity/events?apiVersion=2022-11-28

Reply:
dont just run till we fail. we will be blocked.